route, node: create stub files for Darwin builds

ianvernon opened the pull request:
Running unit tests on Mac OS environments for many packages in Cilium has not
been possible due to the fact that there are dependencies which depend on
variables that are defined only for Linux architectures, e.g. netlink.FAMILY_V6.

Refactor the node and route packages to have Linux-specific and Darwin-specific
files, by using build constraints. For Darwin builds, stubs are created, while
the implementation used by the cilium-agent is factored out into files built
for non-Darwin builds. This allows for running of unit tests on the Mac for
`pkg/policy`, for instance.

No functional change is intended by this.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6306)
<!-- Reviewable:end -->

Commit:
route, node: create stub files for Darwin builds

Running unit tests on Mac OS environments for many packages in Cilium has not
been possible due to the fact that there are dependencies which depend on
variables that are defined only for Linux architectures, e.g. netlink.FAMILY_V6.

Refactor the node and route packages to have Linux-specific and Darwin-specific
files, by using build constraints. For Darwin builds, stubs are created, while
the implementation used by the cilium-agent is factored out into files built
for non-Darwin builds. This allows for running of unit tests on the Mac for
`pkg/policy`, for instance.

No functional change is intended by this.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/datapath/route/route_linux.go:
```diff
@@ -11,6 +11,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
+//
+// +build !darwin
```
```suggestion
// +build linux
```

ianvernon replied:
Fixed in latest upload.

joestringer commented on pkg/datapath/route/route_darwin.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// ByMask is used to sort an array of routes by mask, narrow first.
+type ByMask []Route
```
Does `ByMask` and related functions need to be platform-specific?

ianvernon replied:
I'll factor this out into a separate file, `sort.go`.

ianvernon replied:
Actually, I'll just put this in a file, `route.go` along with the declaration of the `Route` type.

ianvernon replied:
Done.

joestringer commented on pkg/datapath/route/route_darwin.go:
```diff
@@ -0,0 +1,88 @@
+	"github.com/vishvananda/netlink"
+)
+
+type Route struct {
```
Is there a way to share the core route structure (and perhaps the `getLogger()` and other platform-independent code) in a common route.go then only have the platform-specific bits in `route_xxx.go`?

ianvernon replied:
Will do! Thanks for the review.

ianvernon replied:
Done.

joestringer reviewed (dismissed):
LGTM at a glance, I think there's a bit more opportunity for shared code.

Commit:
route: move platform-agnostic route code into route.go

This is shared across Darwin and Linux builds.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
route: change build constraint for route_linux.go

Change from \!darwin to linux.

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on pkg/datapath/route/route_linux.go:
```diff
@@ -0,0 +1,251 @@
+	if err := deleteRoute(route); err != nil {
+		route.getLogger().WithError(err).Error("Unable to delete route")
+		return err
+	} else {
```
Won't do, this was existing logic which I haven't changed.

aanm reviewed (commented):
If we have duplicated files with the same functions but different implementations what are the unit tests testing the darwin environments? Also, wouldn't be easier to just add `+build amd64,linux` for all unit tests that don't work on darwin? I think it would be better than having duplicated files for every architecture we want to run unit tests with.

jrajahalme reviewed (approved):
So the intention is to allow for easier development cycle for non-linux specific packages with unit testing? Looks good to me :-)
